### DISCUSSION TITLE
Localhost Web UI

### DISCUSSION THREAD

**Discussion by lescuer97** [comit-network/xmr-btc-swap#654] [General]
Making a small Localhost Web UI to process the transactions directly wouldn't be a bad idea.  This way the user wouldn't have to copy paste at the end. 
I would totally be glad to develop it, obviously with a little coaching because I'm still not that fluent with Rust.

**Comment by thomaseizinger** [comment]
Are you talking about a Web UI for the user of the CLI?

**Comment by lescuer97** [comment]
Yes for the CLI, it could be simmilar to https://unstoppableswap.net/ but with the functionality of making the swap and seeing the state of the Swaps, keeping everything on the localhost so is a little more secure, or even give it access to the local network so you can check this on the phone. 

I think it would be a great addition for ease of use, maybe for a version 1.0.

**Comment by thomaseizinger** [comment]
GUIs are out of scope from our side.

The CLI can relatively easy be controlled from another process which makes it easy to build a GUI on top without having to modify the source code.

There is also the possibility of building a GUI in Rust and hooking directly into the library functions of the tool. That approach gives you more power and control but is also harder to implement and maintain.

I am going to close this as a wont-fix. If you are interested in building a GUI, I am happy to discuss these options more and provide some guidance.

**Comment by lescuer97** [comment]
Yes, I would totally like to implement a GUI for this and the guidance would be much appreciated.

**Comment by thomaseizinger** [comment]
The first and probably preferable way of building a GUI is to talk to the CLI programmatically.

The idea is simple: write a program with a GUI in whatever programming-language you want. Gather all the input necessary from the user, in the simplest case, that is one input field per command-line option of the CLI.

Once the user wants to swap, you launch the CLI as a background process with the `--json` flag enabled. That will give you structured log output that you can parse. The CLI is non-interactive, i.e. `stderr` will be a stream you can constantly read from, `stdin` can safely be ignored.

The CLI exits automatically once the swap is done.

You can also build some automation around automatically finding providers by using the `list-sellers` command.

**Comment by thomaseizinger** [comment]
Another approach is to directly integrate with the inner workings of the CLI. For this, you will be limited to writing your GUI in Rust because you need to depend on all the code as a library.

This approach gives you a lot more flexibility. For example, you will have direct control over the embedded Bitcoin wallet which allows you to build functionality to manage funds for the user beyond doing a single swap.

If you want to go down that route, your best bet is likely going to be tauri: https://github.com/tauri-apps/tauri

**Comment by lescuer97** [comment]
Thank you for the Input!. I have  been actually looking at writing the GUI in Rust but the GUI environment for GUIs looks very much like a sh*t show, do you have any recommendation other than tauri?

**Reply by thomaseizinger** [reply]
I don't have much experience with any of them but overall, I quite like the direction that [druid](https://github.com/linebender/druid) is going in!

**Comment by lescuer97** [comment]
Hello I have decided to use svelte with Tauri thanks to your recommendation, I have been digging in the swap project, I'm not really clear on how to get a list sellers do I need to do something similar to the  Command::ListSellers  inside the swap.rs file?

**Reply by thomaseizinger** [reply]
I hope you are aware that you've chosen the extra hard path of doing this :)

Listing sellers basically means connecting to the rendezvous node and doing a `DISCOVER` in the relevant namespace, followed by requesting a quote from each seller.

I suggest the following:

- Checkout the libp2p rendezvous specification to understand how the protocol works (it is quite simple, don't worry)
- Familiarize yourself with libp2p, in particular the rust-libp2p implementation which we are using.
- Understand the event loop design pattern that is in place here. Checkout the "file-sharing" example in the rust-libp2p repository.

If you build on top of the swap library and not the CLI then you should only have one event loop in total and not one per CLI command. So in a way yes, you will have to do something similar like `Command::ListSellers` but also quite different because you don't want to terminate after you are done but just report the sellers to the UI, probably as you discover them one by one.

**Reply by lescuer97** [reply]
Yeah hahaahahaha, I'm just noticing the challenge that I picked. I'll be checking the libp2p that you talked. I'm sure to get back with more questions ;).
